pkg/downwardmetrics/virtio-serial/server.go

```diff
@@ -0,0 +1,279 @@
+/*
+ * This file is part of the kubevirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package virtio_serial
+
+import (
+	"bufio"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/textproto"
+	"net/url"
+	"strings"
+	"syscall"
+	"time"
+
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/downwardmetrics"
+	"kubevirt.io/kubevirt/pkg/downwardmetrics/vhostmd/api"
+	diskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
+	metricsScraper "kubevirt.io/kubevirt/pkg/monitoring/domainstats/downwardmetrics"
+	"kubevirt.io/kubevirt/pkg/util"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+)
+
+const (
+	maxConnectAttempts         = 6
+	standardLauncherSocketPath = "/run/kubevirt/sockets/launcher-sock"
+	invalidRequest             = "INVALID REQUEST\n\n"
+	emptyMetrics               = "<metrics><!-- host metrics not available --><!-- VM metrics not available --></metrics>"
+)
+
+func RunDownwardMetricsVirtioServer(nodeName string, stop chan struct{}) error {
+	report, err := newMetricsReporter(nodeName)
+	if err != nil {
+		return err
+	}
+
+	server := downwardMetricsServer{
+		maxConnectAttempts: maxConnectAttempts,
+		virtioSerialSocket: downwardmetrics.DownwardMetricsChannelSocket,
+		reportFn:           report,
+	}
+	return server.start(stop)
+}
+
+type metricsReporter func() (*api.Metrics, error)
+
+func newMetricsReporter(nodeName string) (metricsReporter, error) {
+	exists, err := diskutils.FileExists(standardLauncherSocketPath)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errors.New("virt-launcher socket not found")
+	}
+
+	isolationDetector := isolation.NewSocketBasedIsolationDetector(util.VirtShareDir)
+	scraper := metricsScraper.NewScraper(nodeName, isolationDetector)
```

Comment:
Are we sure that this is now reporting the correct data?
Previously, this would collect information from the host's perspective and now it will collect it from the perspective of the Pod, which is limited to see limited resources and in many cases would not have enough privileges to access host resources. 
This could become even more obvious when the container has limits (for example when the VMI is running with dedicated CPUs)

Are we 100% sure that the data that is reported here is correct?

Reply:
IIUC the isolation detector is just used to get the PID/path of the metrics IO disk like this `vhostmd.NewMetricsIODisk(downwardmetrics.FormatDownwardMetricPath(res.Pid()))`. 

The actual stats are queried from virt-launcher ([see here](https://github.com/kubevirt/kubevirt/blob/84e4ef6c57c09b4b5a0cc7962bbb3d9172d8854d/pkg/virt-handler/cmd-client/client.go#L599)), so that shouldn't be an issue.

Reply:
Besides the doamain stats that we get via libvirtd, so it's the same view inside/outside the container, the rest comes from /proc/{cpuinfo, meminfo, vmstat, stat} and AFAIK these are not cgroup aware(*)

(*) unless the runtime uses something like lxcfs to fake it, but we tested it, and it does not seem to be the case